FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,124 @@
+
+## 1. Data Model
+
+Indexes are used in Xarray to extract data from Xarray objects using coordinate labels instead of using integer array indices. Although the indexes used in an Xarray object can be accessed (or built on-the-fly) via public methods like `to_index()` or properties like `indexes`, those are mainly used internally.
+
+The goal of this project is to make those indexes 1st-class citizens of Xarray's data model. As such, indexes should clearly be separated from Xarray coordinates with the following relationships:
+
+- Index -> Coordinate: one-to-many
+- Coordinate -> Index: one-to-zero-or-one
+
+An index may be built from one or more coordinates. However, each coordinate must relate to one index at most. Additionally, a coordinate may not be tied to any index.
+
+The order in which multiple coordinates relate to an index should matter. For example, Scikit-Learn's `BallTree` index with the Haversine metric requires providing latitude and longitude values in that specific order. As another example, the order in which levels are defined in a `pandas.MultiIndex` may affect its lexsort depth.
+
+Xarray's current data model is already based on the same index-coordinate relationships. The current data model also implies a one-to-one relationship between a dimension and an index. This one-to-one relationship works as currently it is not possible to perform label-based data selection in Xarray using multi-dimensional coordinates, but since we want to enable this feature in this proposal, we need the dimension-index relationship to evolve towards many-to-many.
+
+In the example below, we'd like to select data points based on their x, y values and/or on their latitude/longitude positions:
+
+```
+>>> da
+<xarray.DataArray (x: 2, y: 2)>
+array([[5.4, 7.8],
+       [6.2, 4.7]])
+Coordinates:
+  * lon      (x, y) float64 10.2 15.2 12.6 17.6
+  * lat      (x, y) float64 40.2 45.6 42.2 47.6
+  * x        (x) float64 200.0 400.0
+  * y        (y) float64 800.0 1e+03
+
+>>> da.sel(x=..., y=..., lon=..., lat=...)
+```
+
+We would need one geographic index for the `lat` and `lon` coordinates and two indexes for the `x` and `y` coordinates, respectively.
+
+## 2. Proposed API changes
+
+### 2.1 Index wrapper classes
+
+Every index that is used to select data from Xarray objects should inherit from a base class, e.g., `XarrayIndex`, that provides some common API. `XarrayIndex` subclasses would generally consist of thin wrappers around existing index classes such as `pandas.Index`, `pandas.MultiIndex`, `scipy.spatial.KDTree`, etc.
+
+There is a variety of features that an xarray index wrapper may or may not support:
+
+- 1-dimensional vs. 2-dimensional vs. n-dimensional coordinate
+- built from a single vs multiple coordinate(s)
+- in-memory vs. out-of-core (dask) index data/coordinates (vs. other array backends)
+- orthogonal vs. vectorized indexing, range-based vs. point-wise selection
```

Comment:
This is for integer based indexing?

Reply:
Also for label-based indexing. For example, [pys2index](https://github.com/benbovy/pys2index) currently only supports point-wise selection (nearest-neighbors), but it could later wrap other features available in `s2geometry` like selecting all points within latitude and/or longitude range(s).